_posts/2019-01-30-year-of-the-dog.md

```diff
@@ -0,0 +1,212 @@
+  [eslint-plugin-graphql](https://github.com/apollographql/eslint-plugin-graphql)
+  to warn us when an app is using a field that has been deprecated.
+
+![store2be developer discovering GraphiQL]({{ site.url }}/images/2018_retrospective/developer_discovering_graphql.jpg)
+_store2be developer discovering GraphiQL (photo © unknown)_
+
+## GraphQL in our Rails API
+
+Adopting GraphQL was in retrospect an easy win in the frontend. It did require
+some changes, but towards a more standard workflow with a wide array of tools
+to make our lives easier. It was not so in the API. GraphQL APIs pose a
+different set of problems and require different solutions, compared to purely
+RESTful APIs. To list a few of the challenges we faced:
+
+- **Code organization**. We naturally started by separating the API we expose
+  to users and the one we use for our internal tools, since we had separate
+  sets of endpoints. It seemed like the easiest way to separate what admins and
+  normal users should see and be able to do, and code-sharing would still be
+  possible. For a variety of reasons, this turned out to be a bad decision and
+  we ended up merging the two APIs after a few months. This is only one example
+  of the difficulties we had with proper code organization and reuse in the
+  GraphQL API. We now have a good set of conventions and abstractions but it
+  took a fair amount of exploration.
+- **Testing**. Some of the abstractions in
+  [graphql-ruby](http://graphql-ruby.org/) are not very friendly to test, so we
+  had to come up with a test-friendly code structure.
+- **Limiting access and visibility**. Nothing particularly challenging here,
+  but this also needs to be learned.
+- **Proper use of more "advanced" features like interfaces**
+- **Avoiding the n+1 query problem** - there is one well-supported,
+  well-documented and straightforward solution,
+  [graphql-batch](https://github.com/Shopify/graphql-batch) (by Shopify), but
+  this is one more thing to learn and implement.
+
+It took us a few months to settle on satisfactory abstractions. Resources like
+the [Shopify GraphQL design
+tutorial](https://github.com/Shopify/graphql-design-tutorial) were helpful. In
+general, having someone with prior knowledge or significant interest/experience
+advocating GraphQL good practices and tools helps.
+
+Now that the abstractions are well defined, we find it easier to onboard
+developers to the API, and we work more with plain ruby classes instead of
+Rails-specific methods in models and controllers.
+
+In short, the change was received enthusiastically by the whole team once the
+initial hurdles were behind us, and we are very happy with it.
+
+## Backend and infrastructure - events, k8s
+
+We started implementing an event log for our platform, for auditing, debugging
+and in the future for decoupling tasks. The primary feature at the moment is a
+timeline in our internal app, but it opens up a lot of perspectives and
```

Comment:
If we want to give it a bit more oomph: `which allows us to get a better overview of how various business processes work,`